realm/realm-library/src/main/java/io/realm/BaseRealm.java

```diff
@@ -282,6 +284,38 @@ public void refresh() {
     }
 
     /**
+     * Blocks the current thread until new changes to the Realm are available or {@link #stopWaitForChange}
+     * is called from another thread.
+     *
+     * @return {@code true} If the Realm was updated to the latest version, {@code false} if it was
+     * cancelled by calling {@link #stopWaitForChange}.
+     * @throws IllegalStateException if attempting to wait within a transaction or a Looper thread.
+     */
+    public boolean waitForChange() {
+        checkIfValid();
+        if (isInTransaction()) {
+            throw new IllegalStateException("Cannot wait for changes inside of a transaction.");
+        }
+        if (Looper.myLooper() != null) {
+            throw new IllegalStateException("Cannot wait for changes inside a Looper thread. Use RealmChangeListeners instead.");
+        }
+        boolean hasChanged = sharedGroupManager.getSharedGroup().waitForChange();
+        if (hasChanged) {
+            sharedGroupManager.advanceRead();
+        }
+        return hasChanged;
+    }
+
+    /**
+     * Makes {@link #waitForChange()} return {@code false} immediately. This method is threadsafe
+     * and should be called from another thread than the one that called {@link #waitForChange}.
+     * Calling {@code #stopWaitForChange()} on a Realm that is not waiting for changes has no effect.
+     */
+    public void stopWaitForChange() {
+        sharedGroupManager.getSharedGroup().stopWaitForChange();
```

Comment:
@cmelchior 

> If it only works _if_ the thread is waiting it means that people need to manually synchronize it anyway to prevent deadlocks (e.g. calling `stopWaitForChange` before `waitForChange` was actually triggered). If core does not support this yet, we could easily implement it using a state boolean

Please review the following note to see if I follow what you seek for.

Disregarding normal use case (i.e. we don't call `stopWaitForChange()`) and core change, we can come to following two scenarios where timing of  `stopWaitForChange()` & `waitForChange()` intertwine.

Suppose there is a class private variable **`AtomicBoolean isStoppedByUser`** in `SharedGroup`.

- Case 1

| Time |          Thread 0          |          Thread 1            |
|:------:|:----------------------------:|:----------------------------:|
|   0   | |`waitForChange()`|
|   1    |`stopWaitForChange()`| |

Before Thread 1 goes to wait, it checks if `isStoppedByUser` is `true`. Since it's not, the thread waits. When `waitForChange()` is halted by calling `stopWaitForChange()` from Thread 0, Thread 1 sets `isStoppedByUser` to `false` and stops waiting.

- Case 2

| Time |          Thread 0          |          Thread 1            |
|:------:|:----------------------------:|:----------------------------:|
|   0    |`stopWaitForChange()`| |
|   1   | |`waitForChange()`|

Thread 1 checks if `isStoppedByUser` is `true`. Since it's is (`stopWaitForChange()` is called first), Thread 1 never waits and just sets `isStoppedByUser` to `false`.

I think it can be done quite quickly. Plus, I'd suggest to add a restriction that `waitForChange()` cannot be called from main thread 1) to avoid [this](https://github.com/realm/realm-java/pull/2386#discussion_r55520166) and 2) prevent people from accidentally freeze UI thread.